File: packages/qnsi/src/storage.ts
/**
 * QNSP Storage — PQC-encrypted object storage with SSE-X. Wraps
 * `apps/storage-service` bucket/object routes, registered at `/storage/v1/buckets/*`
 * (apps/storage-service/src/routes/bucket-objects.ts) and reached via the edge gateway's
 * `/proxy/storage` prefix (rewritten to `/storage`).
 */

import type { Internal, RequestOptions } from "./_internal.js";
import { QnsiApiError } from "./errors.js";

// Edge gateway rewrites /proxy/storage -> /storage, so this resolves to the backend's
// /storage/v1/buckets/* routes. (Was /proxy/storage/storage/v1 — a double "storage" that
// rewrote to /storage/storage/v1, which the backend never served, so every call 404'd.)
const PATH_PREFIX = "/proxy/storage/v1";

export interface PutObjectInput {
	readonly data: Uint8Array;
	readonly contentType?: string;
	readonly sseAlgorithm?: string;
	readonly metadata?: Record<string, unknown>;
}

interface GetObjectResponse {
	readonly dataB64?: string;
	readonly [key: string]: unknown;
}

export class StorageClient {
	constructor(private readonly internal: Internal) {}

	putObject(
		bucket: string,
		key: string,
		input: PutObjectInput,
		opts?: Pick<RequestOptions, "idempotencyKey">,
	) {
		const body: Record<string, unknown> = {
			dataB64: encodeB64(input.data),
		};
		if (input.contentType !== undefined) body["contentType"] = input.contentType;
		if (input.sseAlgorithm !== undefined) body["sseAlgorithm"] = input.sseAlgorithm;
		if (input.metadata !== undefined) body["metadata"] = input.metadata;
		return this.internal.request(
			"PUT",
			`${PATH_PREFIX}/buckets/${bucket}/objects/${key}`,
			body,
			opts,
		);
	}

	/**
	 * Returns `[plaintext bytes, descriptor JSON]`. Throws `QnsiApiError`
	 * if the descriptor is missing the dataB64 field.
	 */
	async getObject(bucket: string, key: string): Promise<readonly [Uint8Array, GetObjectResponse]> {
		const resp = await this.internal.request<GetObjectResponse>(
			"GET",
			`${PATH_PREFIX}/buckets/${bucket}/objects/${key}`,
		);
		if (!resp.dataB64) {
			throw new QnsiApiError("storage.getObject: response missing dataB64", 200);
		}
		return [decodeB64(resp.dataB64), resp];
	}

	async deleteObject(bucket: string, key: string): Promise<void> {
		await this.internal.request("DELETE", `${PATH_PREFIX}/buckets/${bucket}/objects/${key}`);
	}

	listObjects(bucket: string, query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/buckets/${bucket}/objects`, undefined, {
			query,
		});
	}

	listBuckets() {
		return this.internal.request("GET", `${PATH_PREFIX}/buckets`);
	}
}

function encodeB64(bytes: Uint8Array): string {
	return Buffer.from(bytes).toString("base64");
}

function decodeB64(b64: string): Uint8Array {
	return new Uint8Array(Buffer.from(b64, "base64"));
}

File: sdks/python/qnsi/src/qnsi/storage.py
"""QNSP Storage - PQC-encrypted object storage with SSE-X."""

from __future__ import annotations

import base64
from typing import Any

from qnsi._errors import QnsiApiError
from qnsi._service import _ServiceClient


class StorageClient(_ServiceClient):
    """Storage client. Wraps ``apps/storage-service`` (``/storage/storage/v1``).

    Bytes are base64-encoded at the boundary; consumers work with raw
    ``bytes`` and the SDK handles the encoding.
    """

    PATH_PREFIX = "/storage/storage/v1"

    def put_object(
        self,
        bucket: str,
        key: str,
        data: bytes,
        *,
        content_type: str | None = None,
        sse_algorithm: str | None = None,
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"dataB64": base64.b64encode(data).decode("ascii")}
        if content_type is not None:
            body["contentType"] = content_type
        if sse_algorithm is not None:
            body["sseAlgorithm"] = sse_algorithm
        if metadata is not None:
            body["metadata"] = metadata
        return self._request(
            "PUT",
            f"/buckets/{bucket}/objects/{key}",
            json=body,
            idempotency_key=idempotency_key,
        )

    def get_object(self, bucket: str, key: str) -> tuple[bytes, dict[str, Any]]:
        """Return ``(plaintext_bytes, descriptor_json)``."""
        resp = self._request("GET", f"/buckets/{bucket}/objects/{key}")
        data_b64 = resp.get("dataB64")
        if not isinstance(data_b64, str):
            raise QnsiApiError(
                "storage.get_object: response missing dataB64",
                status_code=200,
                code=None,
                body=resp,
            )
        return base64.b64decode(data_b64), resp

    def delete_object(self, bucket: str, key: str) -> None:
        self._request("DELETE", f"/buckets/{bucket}/objects/{key}")

    def list_objects(self, bucket: str, **query: Any) -> dict[str, Any]:
        return self._request("GET", f"/buckets/{bucket}/objects", query=query or None)

    def list_buckets(self) -> dict[str, Any]:
        return self._request("GET", "/buckets")
